File: src/main/java/net/ndrei/teslacorelib/inventory/ColoredItemHandler.java
package net.ndrei.teslacorelib.inventory;

import com.google.common.collect.Lists;
import net.minecraft.inventory.Slot;
import net.minecraft.item.EnumDyeColor;
import net.minecraft.item.ItemStack;
import net.minecraftforge.items.IItemHandler;
import net.ndrei.teslacorelib.containers.BasicTeslaContainer;
import net.ndrei.teslacorelib.containers.IContainerSlotsProvider;
import net.ndrei.teslacorelib.gui.BasicTeslaGuiContainer;
import net.ndrei.teslacorelib.gui.IGuiContainerPiece;
import net.ndrei.teslacorelib.gui.IGuiContainerPiecesProvider;

import java.util.List;

/**
 * Created by CF on 2016-12-17.
 */
public class ColoredItemHandler extends FilteredItemHandler implements IContainerSlotsProvider, IGuiContainerPiecesProvider {
    private EnumDyeColor color;
    private String name;
    private BoundingRectangle boundingBox;
    private IFilteredItemHandler containerItemHandler = null;

    public ColoredItemHandler(IItemHandler handler, EnumDyeColor color, String name, BoundingRectangle boundingBox) {
        super(handler);

        this.color = color;
        this.name = name;
        this.boundingBox = boundingBox;
    }

    public EnumDyeColor getColor() {
        return this.color;
    }

    public String getName() {
        return this.name;
    }

    public BoundingRectangle getBoundingBox() { return this.boundingBox; }

//    public final ColoredItemHandlerInfo getInfo() {
//        return new ColoredItemHandlerInfo(this.getName(), this.getColor());
//    }

    @Override
    public List<Slot> getSlots(BasicTeslaContainer container) {
        return Lists.newArrayList();
    }

    @Override
    public List<IGuiContainerPiece> getGuiContainerPieces(BasicTeslaGuiContainer container) {
        return Lists.newArrayList();
    }

    protected IFilteredItemHandler getItemHandlerForContainer() {
        if (this.containerItemHandler == null) {
            this.containerItemHandler = new FilteredItemHandler(this.handler) {
                @Override
                public boolean canInsertItem(int slot, ItemStack stack) {
                    return ColoredItemHandler.this.canInsertItem(slot, stack);
                }
            };
        }
        return this.containerItemHandler;
    }
}

File: src/main/kotlin/net/ndrei/teslacorelib/inventory/ColoredItemHandler.kt
package net.ndrei.teslacorelib.inventory

import net.minecraft.inventory.Slot
import net.minecraft.item.EnumDyeColor
import net.minecraft.item.ItemStack
import net.minecraftforge.items.IItemHandler
import net.ndrei.teslacorelib.containers.BasicTeslaContainer
import net.ndrei.teslacorelib.containers.FilteredSlot
import net.ndrei.teslacorelib.containers.IContainerSlotsProvider
import net.ndrei.teslacorelib.gui.*

/**
 * Created by CF on 2017-06-28.
 */
open class ColoredItemHandler(handler: IItemHandler, val color: EnumDyeColor, val name: String, val index: Int?, val boundingBox: BoundingRectangle)
    : FilteredItemHandler(handler), IContainerSlotsProvider, IGuiContainerPiecesProvider {

    constructor(handler: IItemHandler, color: EnumDyeColor, name: String, boundingBox: BoundingRectangle)
        : this(handler, color, name, null, boundingBox)

    private var containerItemHandler: IFilteredItemHandler? = null

    override fun getSlots(container: BasicTeslaContainer<*>): MutableList<Slot> {
        val result = mutableListOf<Slot>()

        val box = this.boundingBox
        if (!box.isEmpty) {
            val columns = box.width / 18
            (0..this.innerHandler.slots - 1).mapTo(result) {
                FilteredSlot(this.itemHandlerForContainer, it,
                        box.left + 1 + (it % columns) * 18,
                        box.top  + 1 + (it / columns) * 18
                )
            }
        }

        return result
    }

    override fun getGuiContainerPieces(container: BasicTeslaGuiContainer<*>): MutableList<IGuiContainerPiece> {
        val result = mutableListOf<IGuiContainerPiece>()

        val box = this.boundingBox
        if ((!box.isEmpty) && (container.teslaContainer != null)) {
            this.getSlots(container.teslaContainer!!).forEachIndexed { index, it ->
                result.add(TiledRenderedGuiPiece(it.xPos - 1, it.yPos - 1,
                        18, 18,1, 1,
                        BasicTeslaGuiContainer.MACHINE_BACKGROUND, 108, 225, this.color))

                if (this.innerHandler is LockableItemHandler) {
                    result.add(object : BasicContainerGuiPiece(it.xPos, it.yPos, 18, 18) {
                        override fun drawMiddleLayer(container: BasicTeslaGuiContainer<*>, guiX: Int, guiY: Int, partialTicks: Float, mouseX: Int, mouseY: Int) {
                            container.drawTexturedModalRect(
                                    guiX + this.left,
                                    guiY + this.top,
                                    if (this@ColoredItemHandler.innerHandler.locked) 163 else 181, 191, 16, 16)
                        }
                    })

                    result.add(
                        object : GhostedItemStackRenderPiece(box.left, box.top) {
                            override val isVisible: Boolean
                                get() = this@ColoredItemHandler.innerHandler.getStackInSlot(index).isEmpty

                            override fun getRenderStack()
                                    = this@ColoredItemHandler.innerHandler.getFilterStack(index)
                        })
                }
            }
        }

        return result
    }

    protected val itemHandlerForContainer: IFilteredItemHandler
        get() {
            if (this.containerItemHandler == null) {
                this.containerItemHandler = object : FilteredItemHandler(this.innerHandler) {
                    override fun canInsertItem(slot: Int, stack: ItemStack): Boolean {
                        return this@ColoredItemHandler.canInsertItem(slot, stack)
                    }
                }
            }
            return this.containerItemHandler!!
        }
}
